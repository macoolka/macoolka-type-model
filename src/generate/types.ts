/**
 * @file
 */
import { MModule } from '../models/Module'
import { MonadFunctionSync } from 'macoolka-app/lib/MonadFunctionSync'
/**
 * Provide a option at build a TS.
 * @desczh
 * 建立TS的选项
 */
export interface TSOptions {
    /**
     * remark
     * @desczh
     * 注释
     */
    header?: Array<string>
    /**
     * import statement
     * @desczh
     * 导入语句
     */
    imports?: Array<string>
    /**
     * The field that have defaultvalue is optinal when True
     * @desczh
     * 为true时，有缺省值的字段为可选
     */
    isInput?: boolean
    /**
     * Print description when true
     * @desczh
     * 是否打印注释
     */
    showDesc?: boolean
    /**
     * Ignore Type names when validation.
     * @desczh
     * 在校验时忽略给定的类型名字
     */
    ignoreTypeNames?: Array<string>
}
/**
 * Print a Module to ts code text
 * @desczh
 * 打印Module到ts代码
 * @since 0.2.0
 */
export interface TSBuild {
    (option?: TSOptions): MonadFunctionSync<MModule,string>
}
