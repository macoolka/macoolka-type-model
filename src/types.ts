/**
 * Monoid Ord
 * @file
 */
import { Monoid, fold } from 'fp-ts/lib/Monoid'
import { Ord, contramap, ordString } from 'fp-ts/lib/Ord'
import { pipe } from 'fp-ts/lib/pipeable'
import { MModule, MField, MInterface, MClass, MTypeAlias, MFunction, MConstant, MExport, MMethod, MParam } from './models/Module'

import * as Module from './models/Module'
import * as InputModule from './models/InputModule'
export {
    Module,
    InputModule
}
/**
 * @since 0.2.0
 */
export const MMoudleMonoid: Monoid<MModule> = {
    concat: (x, y) => ({
        ...y,
        interfaces: [
            ...x.interfaces ? x.interfaces : [],
            ...y.interfaces ? y.interfaces : []
        ],
        typealiases: [
            ...x.typealiases ? x.typealiases : [],
            ...y.typealiases ? y.typealiases : []
        ],
        classes: [
            ...x.classes ? x.classes : [],
            ...y.classes ? y.classes : []
        ],
        functions: [
            ...x.functions ? x.functions : [],
            ...y.functions ? y.functions : []
        ],
        constants: [
            ...x.constants ? x.constants : [],
            ...y.constants ? y.constants : []
        ],
        exports: [
            ...x.exports ? x.exports : [],
            ...y.exports ? y.exports : []
        ],
    }),
    empty: {
        _kind:'module',
        name: 'macoolka',
        examples: [],
        deprecated: false,
        ignore: false,
        reason: [],
        description: [],
        file: false,
        path: [],
        interfaces: [],
        classes: [],
        typealiases: [],
        functions: [],
        constants: [],
        exports: [],
        idUnique: false,
    }
}
/**
 * @since 0.2.0
 */
export const foldSchema = fold(MMoudleMonoid)


const ordByName = <A extends { name: string }>(): Ord<A> => pipe(
    ordString,
    contramap((value: A) => (value.name))
)
/**
 * @since 0.2.0
 */
export const ordModule = ordByName<MModule>();
/**
 * @since 0.2.0
 */
export const ordField = ordByName<MField>();
/**
 * @since 0.2.0
 */
export const ordInterface = ordByName<MInterface>();
/**
 * @since 0.2.0
 */
export const ordClass = ordByName<MClass>();
/**
 * @since 0.2.0
 */
export const ordFunction = ordByName<MFunction>();
/**
 * @since 0.2.0
 */
export const ordConstant = ordByName<MConstant>();
/**
 * @since 0.2.0
 */
export const ordExport = ordByName<MExport>();
/**
 * @since 0.2.0
 */
export const ordTypeAlias = ordByName<MTypeAlias>();
/**
 * @since 0.2.0
 */
export const ordMethod = ordByName<MMethod>();
/**
 * @since 0.2.0
 */
export const ordParam = ordByName<MParam>();
