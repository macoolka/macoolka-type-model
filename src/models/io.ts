/**
 * Code generated by github.com/macoolka/macoolka-gen-model, DO NOT EDIT.
 */
import * as t from 'macoolka-io'
export const MBasicScalar = t.keyof({"string":"","number":"","boolean":"","int":"","datetime":"","json":""})
export const MStringFormat = t.keyof({"UUID":"","EMail":"","IPV4":"","IPV6":"","URL":""})
export const MIdentifierKind = t.keyof({"field":"","typealias":"","interface":"","class":"","export":"","function":"","param":"","constant":"","module":"","method":"","staticmethod":""})
export const MNameable =t.type({
  name:t.stringMaxLength(64),
})
 
export const MIdentifier =t.type({
  name:MIdentifierKind,
})
 
export const MDocumentable =t.intersection([  t.type({
    /**
     * That this is no longer the preferred way.
     */
    deprecated:t.withDefault(t.boolean,false),
    description:t.withDefault(t.array(t.string),[]),
    ignore:t.withDefault(t.boolean,false),
    examples:t.withDefault(t.array(t.string),[]),
    /**
     * That reason of deprecation.
     */
    reason:t.withDefault(t.array(t.string),[]),
    path:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    descriptions:t.record(t.string,t.any),
    /**
     * When was this feature added.
     */
    since:t.string,
    signature:t.string,
  })
])

export const MExportable =t.type({
  /**
   * That this is no longer the preferred way.
   */
  isExported:t.withDefault(t.boolean,true),
})
 
export const MScalable =t.type({
  /**
   * The Node can be marked as undefined. 
   */
  maybe:t.withDefault(t.boolean,false),
  /**
   * The field has the many multiplicity will also be marked.
   */
  isArray:t.withDefault(t.boolean,false),
  /**
   * The means that no item in the list can be null
   */
  isArrayRequired:t.withDefault(t.boolean,false),
  /**
   * The means that defaultValue is []
   */
  defaultEmptyArray:t.withDefault(t.boolean,true),
})
 
/**
 * Like a Boolean an Enum can have one of a predefined set of values.
 * The difference is that you can define the possible values
 * (whereas for a Boolean the options are restriced to true and false).
 * For example you could specify how an article should be formatted
 * by creating an Enum with the possible values COMPACT, WIDE and COVER.
 */
export const MEnumScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('enum'),
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.string,
  })
])
],'MEnumScalar')

export const MJsonScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('json'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.record(t.string,t.any),
  })
])
],'MJsonScalar')

/**
 * A String holds text. This is the type you would use for a username,
 * the content of a blog post or anything else that is best represented as text.
 */
export const MStringScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('string'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.string,
    minLength:t.int,
    maxLength:t.int,
    pattern:t.string,
    format:MStringFormat,
  })
])
],'MStringScalar')

/**
 * A Number is a number that can have decimals.
 * Use this to store values such as the price of an item in a store or the result of complex calculations.
 */
export const MNumberScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('number'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.number,
    maximum:t.number,
    minimum:t.number,
  })
])
],'MNumberScalar')

/**
 * An Int is a number that cannot have decimals. 
 * Use this to store values such as the weight of an ingredient required for a recipe or the minimum age for an event.
 */
export const MIntScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('int'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.number,
    maximum:t.int,
    minimum:t.int,
  })
])
],'MIntScalar')

/**
 * A Boolean can have the value true or false.
 * This is useful to keep track of settings such as whether the user wants to receive an email newsletter
 * or if a recipe is appropriate for vegetarians.
 */
export const MBooleanScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('boolean'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     */
    defaultValue:t.boolean,
  })
])
],'MBooleanScalar')

/**
 * The DateTime type can be used to store date and/or time values.
 * A good example might be a person's date of birth or the time/data when a specific event is happening.
 */
export const MDateTimeScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('datetime'),
  })
,  t.partial({
    /**
     * You can set a default value for non-list scalar fields.
     * The value will be applied to newly created records when no value was supplied during the create-operation.
     * Now mean current time.
     */
    defaultValue:t.literal('now'),
  })
])
],'MDateTimeScalar')

/**
 * The _kind is a tag that check which Sclar be used.
 */
export const MKindScalar =t.intersection([MScalable,t.intersection([  t.type({
    /**
     * The _kind is a tag that check which Sclar be used.
     */
    _kind:t.literal('kind'),
    value:t.string,
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MKindScalar')

/**
 * A type give a value of model types.
 */
export const MTypeScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('type'),
    /**
     * Model type name.
     */
    value:t.string,
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeScalar')

/**
 * A union type describes a value that can be one of several types.
 */
export const MTypeUnionScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('typeUnion'),
    /**
     * Model type name.
     */
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeUnionScalar')

/**
 * A intersection type describes a value that can be one of several types.
 */
export const MTypeIntersectionScalar =t.intersection([MScalable,t.intersection([  t.type({
    _kind:t.literal('typeIntersection'),
    /**
     * Model type name.
     */
    values:t.withDefault(t.array(t.string),[]),
  })
,  t.partial({
    defaultValue:t.string,
  })
])
],'MTypeIntersectionScalar')

export const MValueable =t.type({
  /**
   *  The type of this Node. A value of type indicates a scalar type.
   */
  type:t.withDefault(t.union([MEnumScalar,MStringScalar,MNumberScalar,MIntScalar,MBooleanScalar,MJsonScalar,MDateTimeScalar,MKindScalar,MTypeScalar,MTypeUnionScalar,MTypeIntersectionScalar,MBasicScalar])
,"string"),
  /**
   * The Node can be marked as required (also referred to as "non-nullable"). 
   */
  required:t.withDefault(t.boolean,false),
})
 
/**
 *  Represents a field in the datamodel.
 * ModelField is referenced by its name and is either scalar or a relation field.
 */
export const MVariable =t.intersection([MDocumentable,MValueable,MNameable,t.type({
})
 ],'MVariable')

/**
 *  Represents a field in the datamodel.
 * ModelField is referenced by its name and is either scalar or a relation field.
 */
export const MField =t.intersection([MVariable,t.type({
  _kind:t.withDefault(t.literal('field'),"field"),
  /**
   * Setting the unique constraint makes sure that two records of the model in question cannot have the same value for a certain field. 
   */
  unique:t.withDefault(t.boolean,false),
  /**
   * Id only one in a model"). 
   */
  id:t.withDefault(t.boolean,false),
  readonly:t.withDefault(t.boolean,false),
  order:t.withDefault(t.boolean,true),
  /**
   * That field exclusive when create operator.
   */
  exclusiveCreate:t.withDefault(t.boolean,false),
  /**
   * That field exclusive when create update.
   */
  exclusiveUpdate:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in where condition.
   */
  exclusiveWhere:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in search result.
   */
  exclusiveSearch:t.withDefault(t.boolean,false),
  /**
   * That field exclusive in load result.
   */
  exclusiveLoad:t.withDefault(t.boolean,false),
})
 ],'MField')

/**
 * ModelParam is referenced by its name and is either scalar or a relation field.
 */
export const MParam =t.intersection([MVariable,t.type({
  _kind:t.withDefault(t.literal('param'),"param"),
})
 ],'MParam')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MFunctional =t.intersection([MDocumentable,MValueable,MNameable,t.type({
  /**
   * Field can be marked as required (also referred to as "non-nullable"). 
   */
  params:t.withDefault(t.array(MParam),[]),
  /**
   * The means that the method return void 
   */
  returnVoid:t.withDefault(t.boolean,false),
})
 ],'MFunctional')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MMethod =t.intersection([MFunctional,t.type({
  _kind:t.withDefault(t.literal('method'),"method"),
})
 ],'MMethod')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MStaticMethod =t.intersection([MFunctional,t.type({
  _kind:t.withDefault(t.literal('staticmethod'),"staticmethod"),
})
 ],'MStaticMethod')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MFunction =t.intersection([MFunctional,MExportable,t.type({
  _kind:t.withDefault(t.literal('function'),"function"),
})
 ],'MFunction')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MConstant =t.intersection([MVariable,MExportable,t.type({
  _kind:t.withDefault(t.literal('constant'),"constant"),
})
 ],'MConstant')

/**
 * Method is referenced by its name and is either scalar or a relation field.
 */
export const MExport =t.intersection([MDocumentable,MNameable,t.type({
  _kind:t.withDefault(t.literal('export'),"export"),
})
 ],'MExport')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MTypeable =t.intersection([MDocumentable,MExportable,MNameable,t.type({
  fields:t.withDefault(t.array(MField),[]),
  methods:t.withDefault(t.array(MMethod),[]),
})
 ],'MTypeable')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MInterface =t.intersection([MTypeable,t.type({
  _kind:t.withDefault(t.literal('interface'),"interface"),
  implements:t.withDefault(t.array(t.string),[]),
})
 ],'MInterface')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MClass =t.intersection([MTypeable,t.type({
  _kind:t.withDefault(t.literal('class'),"class"),
  implements:t.withDefault(t.array(t.string),[]),
  staticMethods:t.withDefault(t.array(MStaticMethod),[]),
})
 ],'MClass')

/**
 * ModelType consist of multiple fields and typically represent entities from your application domain (e.g. User, Car, Order). 
 * An Model Type defines the structure for one model in your datamodel.
 * It is used to represent entities from your application domain.
 */
export const MTypeAlias =t.intersection([MTypeable,MValueable,t.type({
  _kind:t.withDefault(t.literal('typealias'),"typealias"),
})
 ],'MTypeAlias')

/**
 * Define the underlying data model schema.
 */
export const MModule =t.intersection([MDocumentable,MNameable,t.type({
  file:t.withDefault(t.boolean,false),
  _kind:t.withDefault(t.literal('module'),"module"),
  interfaces:t.withDefault(t.array(MInterface),[]),
  typealiases:t.withDefault(t.array(MTypeAlias),[]),
  classes:t.withDefault(t.array(MClass),[]),
  functions:t.withDefault(t.array(MFunction),[]),
  exports:t.withDefault(t.array(MExport),[]),
  constants:t.withDefault(t.array(MConstant),[]),
  idUnique:t.withDefault(t.boolean,false),
})
 ],'MModule')



