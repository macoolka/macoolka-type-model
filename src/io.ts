/**
 * Runtime type system for IO decoding/encoding
 * @desczh
 * 运行时解码/编码系统
 * @file
 */
import { MModule as InputModule } from './models/io'
import * as t from 'macoolka-io'
import { validateFieldName, validateModelName, validateId } from './generate/validation'
import * as A from 'fp-ts/lib/Array'
import { pipe } from 'fp-ts/lib/pipeable'
import { foreach } from './helper'
export * from 'macoolka-io'
import { MModule as IModule } from './models/Module'
export * from './models/io'
import { I18NOption, MonidI18N } from 'macoolka-i18n'
import { defaultOption, foldI18N } from './i18n'
import { merge } from 'macoolka-object'
import * as E from 'fp-ts/lib/Either'
import * as MN from 'macoolka-app/lib/MonadNodeSync'

/**
 * Module IO
 * @since 0.2.0
 *
 */
export const Module = (ignoreNames: Array<string>) => new t.Type<IModule, IModule, unknown>(
    'Module',
    (a): a is IModule => InputModule.is(a),
    (b: any, c) => pipe(
        InputModule.validate(b, c),
        E.chain((u: any) => {
            const result: Array<MN.MonadNodeSync<any>> = []

            foreach(u, {
                model: (model) => {
                    if (u.idUnique === true) {
                        result.push(validateId(model))
                    }
                    result.push(validateFieldName(model))
                },
                schema: (schema) => {

                    result.push(validateModelName(ignoreNames)(schema))

                }
            })

            return pipe(
                result,
                A.lefts,

                as => as.length === 0
                    ? t.success(u)
                    : t.failMessage(foldI18N(as), c)
            )
        })
    )
    ,
    a => a
)
/**
 * Module IO
 * @since 0.2.0
 *
 */
export const NoValidNameModule = new t.Type<IModule, IModule, unknown>(
    'Module',
    (a): a is IModule => InputModule.is(a),
    (b: any, c) => pipe(
        InputModule.validate(b, c),
        E.chain((u: any) => {
            const result: Array<MN.MonadNodeSync<any>> = []

            foreach(u, {
                model: (model) => {
                    if (u.idUnique === true) {
                        result.push(validateId(model))
                    }
                    result.push(validateFieldName(model))
                }

            })

            return pipe(
                result,
                A.lefts,

                as => as.length === 0
                    ? t.success(u)
                    : t.failMessage(foldI18N(as), c)
            )
        })
    )
    ,
    a => a
)
const show = (errors: t.Errors) => (o?: I18NOption) => {
    return t.show(errors)({ i18n: merge(defaultOption, o) })
}
/**
 * Convart Validation to MonadNodeSync
 * @since 0.2.0
 */
export const mapI18N = <A>(value: t.Validation<A>): MN.MonadNodeSync<A> => {
    return pipe(
        value,
        E.mapLeft(a => {
            const m: MonidI18N = (o) => {
                return show(a)(o.i18n)
            }
            return m
        })
    )
}
