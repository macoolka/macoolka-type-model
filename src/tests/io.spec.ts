import * as t from '../io'
import { pipe } from 'fp-ts/lib/pipeable'
import { InputModule as I } from '../types'
import * as E from 'fp-ts/lib/Either'
import BasicA from './fixtures/BasicA'
type InputModule=I.MModule
const en={i18n:{locale:'en'}}
const zh={i18n:{locale:'zh'}}
describe('io', () => {
    it('decode ok', () => {
        
        pipe(
            t.Module([]).decode(BasicA),
            t.mapI18N,
            E.mapLeft(a=>{
                console.log(a({}))
            }),
            result => expect(E.isRight(result)).toBeTruthy()
        )
    })
    describe('validateId', () => {
        it('validateId ok', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',
                        id: true
                    }, {
                        name: 'a2',
                    }, {
                        name: 'a3',
                    }]
                }]
            }
            pipe(
                t.Module([]).decode(schema),
                result => expect(E.isRight(result)).toBeTruthy()
            )
        })
        it('validateId repeat', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',
                        id: true
                    }, {
                        name: 'a2',
                        id: true
                    }, {
                        name: 'a3',
                    }]
                }]

            }
            pipe(
                t.Module([]).decode({ ...schema, idUnique: false }),
                result => expect(E.isRight(result)).toEqual(true)
            )
            pipe(
                t.Module([]).decode(schema),
                t.mapI18N,
                result => {
                    expect(E.isLeft(result)).toEqual(true)
                    return result;
                },
                E.mapLeft(a => {
                    expect(a(en)).toMatchSnapshot()
                    expect(a(zh)).toMatchSnapshot()
                })

            )
           

        })
        it('validateId empty', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',

                    }, {
                        name: 'a2',
                    }, {
                        name: 'a3',
                    }]
                }]

            }
            pipe(
                t.Module([]).decode(schema),
                t.mapI18N,
                E.mapLeft(as => {
                    expect(as(en)).toMatchSnapshot()
                    expect(as(zh)).toMatchSnapshot()
                })
            )
          
        })
    })
    describe('validate field name', () => {
        it('ok', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',
                        id: true
                    }, {
                        name: 'a2',
                    }, {
                        name: 'a3',
                    }]
                }]

            }
            pipe(
                t.Module([]).decode(schema),
                result => expect(E.isRight(result)).toBeTruthy()
            )

        })
        it('validate field name repeat', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test', fields: [{
                        name: 'a1',
                        id: true
                    }, {
                        name: 'a1',

                    }, {
                        name: 'a3',
                    }]
                }]

            }
            pipe(
                t.Module([]).decode(schema),
                t.mapI18N,
                E.mapLeft(as => {
                    expect(as(en)).toMatchSnapshot()
                    expect(as(zh)).toMatchSnapshot()
                })
            )
           

        })

    })

    describe('validate type name', () => {
        it('ok', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',
                        id: true
                    }]
                }, {
                    name: 'test1',
                    fields: [{
                        name: 'a1',
                        id: true
                    }]
                }, {
                    name: 'test2',
                    fields: [{
                        name: 'a1',
                        id: true
                    }]
                }],
                typealiases: [{
                    name: 'E1',

                }, {
                    name: 'E2',
                }]

            }
            pipe(
                t.Module([]).decode(schema),

                result => expect(E.isRight(result)).toBeTruthy()
            )

        })


    })

    describe('validate default value', () => {
        it('ok', () => {
            const schema: InputModule = {
                name: 'schema',
                idUnique: true,
                interfaces: [{
                    name: 'test',
                    fields: [{
                        name: 'a1',
                        type: {
                            _kind: 'string',
                            defaultValue: 'a1'
                        },
                        id: true
                    }, {
                        name: 'a2',
                        type: {
                            _kind: 'int',
                            defaultValue: 1
                        }

                    }, {
                        name: 'a3',
                        type: {
                            _kind: 'number',
                            defaultValue: 2
                        }

                    }, {
                        name: 'a4',
                        type: {
                            _kind: 'datetime',
                            defaultValue: 'now'
                        }

                    }]
                }]

            }
            pipe(
                t.Module([]).decode(schema),

                result => expect(E.isRight(result)).toBeTruthy()
            )

        })


    })
    it('interface empty', () => {
        pipe(
            {},
            t.MInterface.decode,
            t.mapI18N,
            result => {
                expect(E.isLeft(result)).toBeTruthy();
                return result;
            },
            E.mapLeft(as => {
                expect(as(en)).toMatchSnapshot()
                expect(as(zh)).toMatchSnapshot()
            })
        )
       
    })
    it('interface ok', () => {
        pipe(
            { name: '1' },
            t.MInterface.decode,
            result => {
                expect(E.isRight(result)).toBeTruthy();

            },

        )

    })
})


