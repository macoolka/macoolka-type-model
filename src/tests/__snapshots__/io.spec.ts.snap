// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io interface empty 1`] = `"Invalid value undefined supplied to : MInterface/0: MTypeable/2: { name: stringMaxLength=64 }/name: string max length=64"`;

exports[`io interface empty 2`] = `"无效的值undefined : MInterface/0: MTypeable/2: { name: stringMaxLength=64 }/name: 最大长度=64"`;

exports[`io validate field name validate field name repeat 1`] = `
"
Field name [\\"\\\\\\"a1\\\\\\"\\"] repeat in Model[test]. : Module"
`;

exports[`io validate field name validate field name repeat 2`] = `
"
在模型(test)中字段名称(\\"\\\\\\"a1\\\\\\"\\")重复. : Module"
`;

exports[`io validateId validateId empty 1`] = `
"
Id not found  in test : Module"
`;

exports[`io validateId validateId empty 2`] = `
"
模块test必须有一个Id字段 : Module"
`;

exports[`io validateId validateId repeat 1`] = `
"
Id[a1,a2] repeat in Model[test]. : Module"
`;

exports[`io validateId validateId repeat 2`] = `
"
Id[a1,a2]在模块[test]中重复. : Module"
`;
